### ISSUE TITLE
Improved language support 

### ISSUE THREAD

**Issue by padde** [lau/calendar#10]
Hi @lau,

I just watched your talk from ElixirConf 2015, where you encouraged us to add translations for more languages. Here are some of my thoughts on how to improve language support. I would love to implement these changes but wanted to make sure we're on the same page before starting a PR. Please let me know what you think regarding the following points:
1. The API should allow users to supply custom translations, because the default translations will certainly not work for everyone. For example in Germany there are at least two ways of abbreviating week day names: `Mo, Di, Mi, Do, Fr, Sa, So` but (very rarely) you also see `Mon, Die, Mit, Don, Fre, Sam, Son`. It would definitely be nice to be able to override parts of the translations, just day names for example.
2. Maybe using [`gettext`](https://github.com/elixir-lang/gettext) would be a good idea, since it seems to become the de-facto standard for translations in Elixir. I am not sure whether it would be possible to allow user-defined gettext files which could extend the existing languages and/or override existing defaults. 
3. If we should decide against an external translation library, I think that the approach of just defining the translations inline in the `Strftime` module will not scale very well since we would clutter the file with all the translations. In that case I would suggest extracting the translations and related functions into a separate module.
4. Others have already done a lot of translation work before us, which we can leverage. For example the rails-i18n project has a large amount of YAML files with default translations for all kinds of languages, which can be found [on Github](https://github.com/svenfuchs/rails-i18n/tree/master/rails/locale). We could easily read those files to generate translations – low hanging fruit.
5. If you carefully look at the rails-i18n translations, you will notice that they have translations for civil date/time formats like `October 25, 2015 01:07 PM`, which could be a nice addition.

I am looking forward to hear what you think!

Cheers
Patrick

**Comment by lau** [comment]
Hi Patrick

It is nice that you offer help. After the talk José Valim advised me to take the translations out of Calendar. I agree with that. It should be a separate library.

The way the translations are structured now is not ideal anyway, it is an interim solution.

Right now I'm thinking that a protocol for translations is a good idea. And then it would be good to have a library/package that implements that protocol and has translations for various languages. Then if someone disagrees with certain identifiers or translations they can use a different translation-package.
A translation library name could be defined in config.exs

Yes, it would be good to use gettext for the translations. And it sounds like a good idea to use existing data from other projects.

Maybe I should make the protocol first and then we can take it from there.

**Comment by padde** [comment]
Hi Lau,

that sounds like a good idea, a translation protocol would certainly be a good first step. Let me know if I can do something to help.

**Comment by lau** [comment]
Hi @padde 

Now with 0.11.1 there is a module `Calendar.DefaultTranslations` (see https://github.com/lau/calendar/blob/master/lib/calendar/strftime.ex#L271 ) The implementation is very simple right now.

I started with a protocol, but realized that it doesn't make sense in this situation. Instead other modules will just have to provide a module with a similar API. Right now the functions are `weekday_names`, `weekday_names_abbr`, `month_names`, `month_names_abbr`. Then in the config.exs file people can use another module name instead of `Calendar.DefaultTranslations`.

Do you want to make a project that provides these translations for more languages?

**Comment by padde** [comment]
Hi @lau,

great news! Unfortunately I'm quite busy at the moment so I don't know when I will find the time to do it. I'd be glad to help but if you can't wait, I'll not be offended if you start the project yourself :wink:

**Comment by padde** [comment]
@lau in fact I did find some time to write a basic script that extracts the translations from rails-i18n. The resulting hex package is here:

https://github.com/padde/calendar_translations

Any suggestions and improvements are most welcome.

**Comment by padde** [comment]
I think this can be closed now, but I'll leave it up to you to decide :wink:

**Comment by padde** [comment]
I think this can be closed now, but I'll leave it up to you to decide :wink:

**Comment by lau** [comment]
Hi @padde. Thanks for the translations package. I haven't looked at this in a while. Maybe I will add a behavior and ping you about it some time.
